### DISCUSSION TITLE
Separate Reporting from Backtest#Run()

### DISCUSSION THREAD

**Discussion by ibyter** [cinar/indicator#210] [Ideas]
I think it would be better if the `Backtest#Run()` method returns the result details as a struct.  Then, there is a separate a method/function e.g.`GenerateHtmlReport`  that would create the .html reports.

In my case, I want to be able to further interact with the backtest result (programmatically), so that I can go through multiple combinations without getting constrained or limited by the html reports.

**Comment by cinar** [comment]
I'm working with a large dataset of assets and strategies, so it will be challenging to scale if I store all the results. However, with the new approach, I can easily create a solution similar to what you've described. Can you tell me what information should be in those structs after the backtest?

**Reply by ibyter** [reply]
Basically two pieces of information:
1. Outcome
2. List of action/time/price

**Reply by cinar** [reply]
Excellent, so probably a mapping like this would work?

```
type StrategyResult struct {
    assetName string
    strategy Strategy
    outcome float64
    action Action
    transactions []Action
}

type  BacktestResult  map[string][]StrategyResult
```

I left the time and price information out, as they are already available, but let me know.

**Comment by ibyter** [comment]
Those will be good to begin with. But regarding `transactions []Action`, is it normailized? because in my case, I want to know that at time T when price was  P, action A was taken. Basically I want to capture the actions, relative to price/time, as illustrated below:
![image](https://github.com/user-attachments/assets/87750431-58a9-435e-b2e9-8217b00a260e)

**Reply by cinar** [reply]
I think it's best to keep the data unnormalized here. That way, we can see the recommended action based on the strategy, without being influenced by past actions. We can normalize it later if needed.

I'll try to put together something and share it here.

**Reply by cinar** [reply]
I tried to put this together. Could you check #218?
Is this similar to what you were thinking?

**Reply by ibyter** [reply]
Thanks, I have taken a look. So is this how it will be used by backtest:

```Go
myreport  := backtest.NewDataReport()
mybacktest := backtest.NewBacktest(repository,  myreport)
mybacktest.Names = append(mybacktest.Names, "somestock")
mybacktest.Strategies = append(mybacktest.Strategies, momentum.NewStochasticRsiStrategy())
err := mybacktest.Run()


// result will be in myreport
fmt.Println("outcome is: ", myreport.Results["somestock"][0].Outcome)
```

**Reply by cinar** [reply]
Yes, that's how it should work. The Results are not ordered currently. I can easily make them ordered by the outcome. Would this work for you?

**Reply by ibyter** [reply]
the current implementation works perfectly fine for me, sorting isn't necessary

**Reply by cinar** [reply]
Excellent, I pulled the changes and produced a new release. You should be able to start using this starting from [v2.1.4](https://github.com/cinar/indicator/releases/tag/v2.1.4).
